compiler/rustc_mir/src/transform/promote_consts.rs

```diff
@@ -597,22 +548,67 @@ impl<'tcx> Validator<'_, 'tcx> {
+        if let BorrowKind::Mut { .. } = kind {
+            let ty = Place::ty_from(place.local, place.projection, self.body, self.tcx).ty;
+
+            // In theory, any zero-sized value could be borrowed
+            // mutably without consequences. However, only &mut []
+            // is allowed right now, and only in functions.
+            if let ty::Array(_, len) = ty.kind() {
+                match len.try_eval_usize(self.tcx, self.param_env) {
+                    Some(0) => {}
+                    _ => return Err(Unpromotable),
+                }
+            } else {
+                return Err(Unpromotable);
+            }
+        }
+
+        // HACK(eddyb) this should compute the same thing as
+        // `<HasMutInterior as Qualif>::in_projection` from
+        // `check_consts::qualifs` but without recursion.
+        let mut has_mut_interior = self.qualif_local::<qualifs::HasMutInterior>(place.local);
+        if has_mut_interior {
+            let mut place_projection = place.projection;
+            // FIXME(eddyb) use a forward loop instead of a reverse one.
+            while let &[ref proj_base @ .., elem] = place_projection {
+                // FIXME(eddyb) this is probably excessive, with
+                // the exception of `union` member accesses.
+                let ty = Place::ty_from(place.local, proj_base, self.body, self.tcx)
+                    .projection_ty(self.tcx, elem)
+                    .ty;
+                if ty.is_freeze(self.tcx.at(DUMMY_SP), self.param_env) {
+                    has_mut_interior = false;
+                    break;
+                }
+
+                place_projection = proj_base;
+            }
+        }
+        if has_mut_interior {
+            return Err(Unpromotable);
+        }
+
+        Ok(())
+    }
+
     fn validate_rvalue(&self, rvalue: &Rvalue<'tcx>) -> Result<(), Unpromotable> {
         match *rvalue {
-            Rvalue::Cast(CastKind::Misc, ref operand, cast_ty) if self.const_kind.is_none() => {
+            Rvalue::Cast(CastKind::Misc, ref operand, cast_ty) if self.maybe_runtime() => {
                 let operand_ty = operand.ty(self.body, self.tcx);
                 let cast_in = CastTy::from_ty(operand_ty).expect("bad input type for cast");
                 let cast_out = CastTy::from_ty(cast_ty).expect("bad output type for cast");
                 match (cast_in, cast_out) {
                     (CastTy::Ptr(_) | CastTy::FnPtr, CastTy::Int(_)) => {
-                        // in normal functions, mark such casts as not promotable
+                        // ptr-to-int casts are not promotable
                         return Err(Unpromotable);
                     }
                     _ => {}
                 }
             }
 
-            Rvalue::BinaryOp(op, ref lhs, _) if self.const_kind.is_none() => {
+            Rvalue::BinaryOp(op, ref lhs, _) if self.maybe_runtime() => {
```

Comment:
Pointer comparison/offset is not stable in any const context. We should never promote these.

Reply:
Also missing here is floating point arithmetic, so we currently promote e.g. `&(42.0 * 55.1)`. This is actually okay, because the IEEE spec mandates the precision for simple arithmetic operations, so there shouldn't be a difference between the run-time and compile-time result. I wonder if this was a conscious choice way way back, or just a happy accident.

Reply:
It probably was an accident.